squirrel-34: + mongo db service for later usage

nazar-kotiukha opened the pull request:
I added the MongoDB service to extend its functionality for future use with entities that are needed in the project. Also, I included configurations for future use and added services to the DI container.

nazar-kotiukha committed:
squirrel-34: + added IMongoService interface

nazar-kotiukha committed:
squirrel-34: + added mongoDb configurations and service

 added mongoDb configurations for future use

Anatolii-Ivanov commented on backend/Squirrel.Core/Squirrel.Core.BLL/Services/MongoService.cs:
Can you make it a bit more generic, so it will be possible to define collection name in DI (As currently it's assigned via constructor and it's impossible to work with 2 collections)

nazar-kotiukha replied:
Of course! Tried to make it generic and pushed it in a new commit. Let me know please if this way is okay.

serhiiyasenev commented on backend/Squirrel.Core/Squirrel.Core.BLL/Services/MongoService.cs:
```diff
@@ -0,0 +1,66 @@
+
+        public async Task<SampleDto> UpdateAsync(int sampleId, NewSampleDto sampleDto)
+        {
+            var existedSample = await _sampleCollection.Find(x => x.Id == sampleId).FirstOrDefaultAsync();
```
can we simplify to use just `FirstOrDefaultAsync` without `Find`?

nazar-kotiukha replied:
IMongoCollection<TDocument> doesn't contain FirstOrDefaultAsync to use it instead of Find, so as far as i concerned Find is a preferable way

nazar-kotiukha committed:
sql-34: * made service generic

now collectionName will be assigned based on typeof T, if type doesn't match any case, exception will be thrown

nazar-kotiukha committed:
sql-34: * fixed errors while injection

removed unnecessary parameter from constructor

nazar-kotiukha commented:
Fixed some issues with injection in the latest commit

serhiiyasenev commented on backend/Squirrel.Core/Squirrel.Core.BLL/Services/MongoService.cs:
```diff
@@ -0,0 +1,62 @@
+
+            var mongoDatabase = mongoClient.GetDatabase(mongoDatabaseSettings.Value.DatabaseName);
+
+            string collectionName = typeof(T) switch
```
probably it's better to provide `collectionName` via constructir instead of getting it from type (`typeof(T) switch`)? 

it can be not very clear (this swith), but I'm not sure

nazar-kotiukha replied:
agree if that, changed logic will commit shortly

serhiiyasenev commented on backend/Squirrel.Core/Squirrel.Core.WebAPI/Extentions/ServiceCollectionExtensions.cs:
```diff
@@ -1,23 +1,29 @@
+            services.Configure<MongoDatabaseConnectionSettings>(
+                configuration.GetSection("MongoDatabase"));
+
+            services.AddScoped(typeof(IMongoService<>), typeof(MongoService<>));
```
why not `services.AddTransient/Scoped<IMongoService, MongoService>();`?

if new on each request then `Scoped`

nazar-kotiukha committed:
sql-34: * changed logic

Updated injection logic, removed unused inheritance from BaseService, and introduced the collectionName parameter to the constructor.

serhiiyasenev commented on backend/Squirrel.Core/Squirrel.Core.WebAPI/Extentions/ServiceCollectionExtensions.cs:
```diff
@@ -1,23 +1,40 @@
+                var context = serviceProvider.GetRequiredService<SquirrelCoreContext>();
+                var mapper = serviceProvider.GetRequiredService<IMapper>();
+                var mongoDatabaseSettings = serviceProvider.GetRequiredService<IOptions<MongoDatabaseConnectionSettings>>();
+                string collectionName = "SampleCollection";
```
can we get `string collectionName = "SampleCollection";` from settings instead of hardcode?

nazar-kotiukha replied:
Initially, I approached it that way, but Anatolii pointed out that using this approach would limit us from using it with more than one collection: "so it will be possible to define collection name in DI (As currently it's assigned via constructor and it's impossible to work with 2 collections)"

serhiiyasenev replied:
so leave it as is?

nazar-kotiukha replied:
I currently believe that it's not such a bad approach. Additionally, this sample collection has been added here for demonstrative purposes. I will explore better ways when MongoDB will actually be used in our project.

serhiiyasenev replied:
yeah, let it be

Zoranais commented on backend/Squirrel.Core/Squirrel.Core.WebAPI/Extentions/ServiceCollectionExtensions.cs:
```diff
@@ -1,23 +1,40 @@
 
             services.AddTransient<ISampleService, SampleService>();
+
+            services.Configure<MongoDatabaseConnectionSettings>(
+                configuration.GetSection("MongoDatabase"));
+
+            services.AddTransient(typeof(IMongoService<Sample>), serviceProvider =>
+            {
+                var context = serviceProvider.GetRequiredService<SquirrelCoreContext>();
+                var mapper = serviceProvider.GetRequiredService<IMapper>();
+                var mongoDatabaseSettings = serviceProvider.GetRequiredService<IOptions<MongoDatabaseConnectionSettings>>();
+                string collectionName = "SampleCollection";
+
+                return new MongoService<Sample>(mongoDatabaseSettings, collectionName);
+            });
```
It would probably be better to move this part to a separate extension method `AddMongoDbService`. Also why are you getting context and mapper here?

nazar-kotiukha replied:
Oh, my bad. I forgot to remove it. Thanks!

Zoranais commented on backend/Squirrel.Core/Squirrel.Core.WebAPI/appsettings.json:
```diff
@@ -2,6 +2,10 @@
   "ConnectionStrings": {
     "SquirrelCoreDBConnection": "Server=localhost;Database=SquirrelCoreDB;Trusted_Connection=True;"
   },
+  "MongoDatabase": {
+    "ConnectionString": "",
+    "DatabaseName": ""
+  },
```
I think, it would probably be better to move this to the `ConnectionStrings` block and get rid of the `MongoDataBaseConnectionSettings` model.

nazar-kotiukha replied:
I agree with the placement suggestion, but what is the benefit of removing MongoDatabaseConnectionSettings?

Anatolii-Ivanov replied:
It's ok to leave it here

nazar-kotiukha committed:
sql-34: * removed unused services

serhiiyasenev commented on backend/Squirrel.Core/Squirrel.Core.BLL/Interfaces/IMongoService.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace Squirrel.Core.BLL.Interfaces
+{
+    public interface IMongoService<T> where T : Entity<long>
```
why `Entity<long>`?

nazar-kotiukha replied:
I used Entity<long> to maintain consistency with our AuditEntity<T> pattern. This approach ensures clarity and coherence throughout codebase. Correct me if I'm wrong, please. Thanks!

serhiiyasenev replied:
I mean why `long`?

nazar-kotiukha replied:
As far as I am concerned C# does not support "generics on generics", so picked long as it was used in Sampe entity class. I think this too will be more clear in future when we will have a full picture of MongoDb usage and I will change it, if needed.

serhiiyasenev replied:
thanks for the explanation!

Anatolii-Ivanov commented on backend/Squirrel.Core/Squirrel.Core.WebAPI/Extentions/ServiceCollectionExtensions.cs:
```diff
@@ -1,23 +1,44 @@
+
+        public static void AddMongoDbService(this IServiceCollection services)
+        {
+            services.AddTransient(typeof(IMongoService<Sample>), serviceProvider =>
+            {
+                var mongoDatabaseSettings = serviceProvider.GetRequiredService<IOptions<MongoDatabaseConnectionSettings>>();
+                string collectionName = "SampleCollection";
+
+                return new MongoService<Sample>(mongoDatabaseSettings, collectionName);
+            });
```
```suggestion
            services.AddTransient<IMongoService<Sample>>(s => 
                new MongoService(s.GetRequiredService<IOptions<MongoDatabaseConnectionSettings>>(), "SampleCollection");
```

nazar-kotiukha committed:
sql-34: * more efficient service registration

Co-authored-by: Anatolii Ivanov <57347711+Anatolii-Ivanov@users.noreply.github.com>

serhiiyasenev commented on backend/Squirrel.Core/Squirrel.Core.BLL/Services/MongoService.cs:
```diff
@@ -0,0 +1,53 @@
+
+        public async Task<T> CreateAsync(T entity)
+        {
+            await _mongoCollection.InsertOneAsync(entity);
```
probably

return await _mongoCollection.InsertOneAsync(entity);

?

nazar-kotiukha replied:
this method doesn't return entity

serhiiyasenev commented on backend/Squirrel.Core/Squirrel.Core.BLL/Services/MongoService.cs:
```diff
@@ -0,0 +1,53 @@
+
+        public async Task<T> UpdateAsync(int entityId, T editedEntity)
+        {
+            await _mongoCollection.ReplaceOneAsync(x => x.Id == entityId, editedEntity);
```
probably

return await _mongoCollection.ReplaceOneAsync(x => x.Id == entityId, editedEntity);

?

nazar-kotiukha replied:
this method doesn't return entity also

nazar-kotiukha committed:
34: * resolved merge conflict

nazar-kotiukha committed:
Revert "34: * resolved merge conflict"

This reverts commit 63f11c359694b31dadd022e6388113eaf5a0b26a, reversing
changes made to 6ccd247849f78f21bc3f4fb0807acf180b92fd26.

revert
